File: 494. Target Sum - Medium/solution.py
class Solution:
    def findTargetSumWays(self, nums: List[int], S: int) -> int:
        if not nums or S > 1000:
            return 0
        
        dic = {nums[0]: 1, -nums[0]: 1} if nums[0] else {0: 2}
        
        for i in range(1, len(nums)):
            tdic = {}
            
            for d in dic:
                tdic[d + nums[i]] = tdic.get(d + nums[i], 0) + dic.get(d, 0)
                tdic[d - nums[i]] = tdic.get(d - nums[i], 0) + dic.get(d, 0)
            
            dic = tdic
        
        return dic.get(S, 0)

File: 494. Target Sum - Medium/solution.cpp
static auto _ = [] () { ios_base::sync_with_stdio(false); cin.tie(nullptr); return 0; } ();
class Solution {
public:
    int findTargetSumWays(vector < int >& nums, int S) {
        if (S > 1000) {
            return 0;
        }
        
        vector < int > dp(2001, 0);

        dp[f(+nums[0])] += 1;
        dp[f(-nums[0])] += 1;
        
        for (int i = 1; i < nums.size(); ++i) {
            vector < int > next(2001, 0);
            
            for (int s = -1000; s <= 1000; ++s) {
                if (dp[f(s)] > 0) {
                    next[f(s+nums[i])] += dp[f(s)];
                    next[f(s-nums[i])] += dp[f(s)];                   
                }
            }
            
            dp = std::move(next);
        }
        
        return dp[f(S)];
    }
    
private:
    inline int f(int x) {
        return x + 1000;
    }
};